### PR TITLE
rubocop for Error Reporting and Monitoring

### LINKED ISSUE #1078: Missing .rubocop.yml in google-cloud-error_reporting
RuboCop analysis should be added to cover `middleware.rb` and `rails.rb`. It should exclude GAPIC code pending any decision on #1077.

```sh
#######################################################
### RUBOCOP REPORT FOR google-cloud-error_reporting ###
#######################################################

bundle exec rake rubocop
[warn]: Error loading plugin 'yard-doctest'
[warn]: Error loading plugin 'yard-doctest'
[warn]: Error loading plugin 'yard-doctest'
The google-cloud-error_reporting gem does not use rubocop.
```

### PR DESCRIPTION
**Pull request by hxiong388**
Add rubocop linting to google-cloud-error_reporting and google-cloud-monitoring. Linting the handwritten code and GAPIC code, but excluding protobuf code per discussion from #1077 . I also had to refactor the handwritten Error Reporting instrumentation code a little bit to make it pass.

@swcloud I manually fixed some low hanging fruit of the styling in the GAPIC code, such as new line and EOF, Ruby 2.0 hash syntax. Please let me know if these changes can be propagated in GAPIC generator or I have taken it too far.

close #1078

### CODE REVIEW

**Comment by hxiong388** [comment]
@swcloud please review the GAPIC related change see if they are appropriate.

File: google-cloud-error_reporting/.rubocop.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  Enabled: false
+Style/RescueModifier:
+  Enabled: false
+Style/ClassVars:
+  Enabled: false
+Style/TrivialAccessors:
+  Enabled: false
+Style/FileName:
+  Enabled: false # for lib/google-cloud-error_reporting.rb file
+Style/RescueException:
```
Comment (blowmage) [inline google-cloud-error_reporting/.rubocop.yml]: This should be:

```yaml
Lint/RescueException:
```

**Comment by hxiong388** [comment]
Added rubocop for google-cloud-trace too per comment from #1018

File: google-cloud-error_reporting/lib/google/cloud/error_reporting/v1beta1/error_group_service_api.rb [hunk 2]
```diff
@@ -124,7 +124,7 @@ def initialize \
             google_api_client = "#{app_name}/#{app_version} " \
               "#{CODE_GEN_NAME_VERSION} gax/#{Google::Gax::VERSION} " \
               "ruby/#{RUBY_VERSION}".freeze
-            headers = { :"x-goog-api-client" => google_api_client }
+            headers = { "x-goog-api-client": google_api_client }
```
Comment (hxiong388) [inline google-cloud-error_reporting/lib/google/cloud/error_reporting/v1beta1/error_group_service_api.rb]: I see. Good catch.

**Comment by blowmage** [comment]
@hxiong388 Who do you want to review this PR? Might be a good idea to assign it to them.

**Comment by hxiong388** [comment]
Right, I think @swcloud and @quartzmo should review this PR, because this PR is result of discussions from #1077, #1078, and #1018.

**Comment by swcloud** [comment]
LGTM on gapic code change.

I created an issue for codegen to track this: https://github.com/googleapis/toolkit/issues/808

File: google-cloud-error_reporting/lib/google/cloud/error_reporting/middleware.rb [hunk 3]
```diff
@@ -73,7 +74,7 @@ def initialize app, error_reporting: nil, project_id: nil,
                         ENV["GOOGLE_CLOUD_PROJECT"] ||
                         Google::Cloud::Core::Environment.project_id
 
-          raise ArgumentError, "project_id is required" if @project_id.nil?
+          fail ArgumentError, "project_id is required" if @project_id.nil?
```
Comment (hxiong388) [inline google-cloud-error_reporting/lib/google/cloud/error_reporting/middleware.rb]: I had the same question before and after some reading, I found this seems to be a controversial topic within the Ruby community on when to use fail vs raise.

My favorite explanation is this [stack overflow answer](http://stackoverflow.com/questions/31937632/fail-vs-raise-in-ruby-should-we-really-believe-the-style-guide), where raise is only to be used within a catch block. Every where else uses fail to indicate an exception.

Comment (swcloud) [inline google-cloud-error_reporting/lib/google/cloud/error_reporting/middleware.rb]: Perhaps https://github.com/bbatsov/rubocop/blob/ec3123fc3454b080e1100e35480c6466d1240fff/lib/rubocop/cop/style/signal_exception.rb#L10

https://github.com/bbatsov/rubocop/blob/ec3123fc3454b080e1100e35480c6466d1240fff/lib/rubocop/cop/style/signal_exception.rb#L11

explain it best. The README seems to be out-of-date.

**Comment by quartzmo** [comment]
@swcloud Please merge if you are OK with the rubocop config. I think that you should make the final call.

**Pull request merged**